Make sure to lock cs_mapBlockIndex when accessing nStatus or nSequenceId

ptschip opened the pull request:
Most of the data in blockindex is immutable and doesn't need to be locked when we access it as long as we have a valid pointer. We only need to lock when we actually do a lookup to the blockindex (which we do already) but since moving to the atomic chainActive.TIp() we are at times not locking the underlying mutable data: nStatus and nSequenceId.  This PR remedies that and locks all access to these two items.

ptschip committed:
Lock nSequenceId with cs_mapBlockIndex

There were some areas that were not locked fully or properly. We have
to lock any blockindex data that is mutable. Currently only nSequenceId
and nStatus need to be locked when we read or update these values.

ptschip committed:
When we access nStatus we must have a lock on  cs_mapBlockIndex

Generally when accessing the underlying data from a blockindex pointer
we don't have to lock anything except for the access to the map in
aquire the pointer. This is because the underlying data is immutable
with the exception of nStatus and nSequenceId.

In this commit we make sure to lock any acesses, either read or write.
to nStatus. Access to nSequenceId were dealt with in the previous
commit.

ptschip committed:
In ProcessGetData always make sure we are using a valid blockindex

It's rather easy to make a mistake in this section of code which
could open up a remote crash attack. This could happen if a developer
failed to use the fSend flag before accessing the blockindex pointer. The fSend
flag in this section of code has been used to test whether the blockindex
is null or not, which is not a good practice.

sickpig commented:
@ptschip 

seems like `schnorrigs.py` is failing 

```
stderr:
 bitcoind: ../../src/sync.cpp:140: void potential_deadlock_detected(const std::pair<void*, void*>&, const LockStack&, const LockStack&): Assertion `onlyMaybeDeadlock' failed.
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 235, in main
    self.run_test()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/schnorrsig.py", line 265, in run_test
    self.nodes[2].invalidateblock(forkingBlocks[0])
```

ptschip committed:
Fix potential deadlock

Locking order shoud be cs_main->cs_wallet->cs_mapBlockIndex but
was set to cs_main->cs_mapBlockIndex. In InvalidateBlock(), exxit
the scope where we take cs_mapBlockIndex before we DisconnectTip
which involves the wallet.

ptschip commented:
thx, fixed...

On 02/05/2019 12:13 p.m., Andrea Suisani wrote:
>
> @ptschip <https://github.com/ptschip>
>
> seems like |schnorrigs.py| is failing
>
> |stderr: bitcoind: ../../src/sync.cpp:140: void
> potential_deadlock_detected(const std::pair<void*, void*>&, const
> LockStack&, const LockStack&): Assertion `onlyMaybeDeadlock' failed.
> File
> "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py",
> line 235, in main self.run_test() File
> "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/schnorrsig.py",
> line 265, in run_test self.nodes[2].invalidateblock(forkingBlocks[0]) |
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1715#issuecomment-488795824>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUDJY2EPEG4JGAVYAOTPTM4MFANCNFSM4HJ776DA>.
>

ptschip committed:
Improve reliability of keypool.py

ptschip committed:
Fix GetChainTips

gandrewstone commented on src/net_processing.cpp:
```diff
@@ -94,144 +94,156 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
-                        }
-                        else if (inv.type == MSG_CMPCT_BLOCK)
+                        READLOCK(cs_mapBlockIndex);
+                        fHaveData = (mi->nStatus & BLOCK_HAVE_DATA);
```
If the system can tolerate fHaveData being incorrect then we don't need the read lock.  If the system can't tolerate it being incorrect then the lock needs to be extended to cover ReadBlockFromDisk.  Otherwise the record could be deleted and mi->nStatus & BLOCK_HAVE_DATA changed after the lock is released but before the record is read from disk.

For performance reasons, it makes sense to allow fHaveData to be incorrect and gracefully degrade functionality (in this case, don't provide the block) in that rare case.  So I don't think that we need the lock here.

ptschip replied:
I think you're right, I think we can tolerate a degraded mode so we can remove this lock.

gandrewstone commented on src/net_processing.cpp:
```diff
@@ -94,144 +94,156 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
-                    CBlock block;
-                    if (!ReadBlockFromDisk(block, mi, consensusParams))
+                        // disconnect node
+                        pfrom->fDisconnect = true;
```
Why disconnect if we end up not having the block?  The problem is on our side

ptschip replied:
I think you might be thinking of something else...this is old code that's been there and disconnects the peer if we've reached the block serving limit that the node operator has configured..I dont' think you see it used much but it was something put in by core to prevent to much IBD peers from using up your outbound bandwidth.

Also this is for processing getdata requests, so we're sending here, not receiving.

gandrewstone commented on src/net_processing.cpp:
```diff
@@ -94,144 +94,156 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
+                            else
+                            {
+                                // BU: don't relay excessive blocks that are not on the active chain
+                                if (mi->nStatus & BLOCK_EXCESSIVE)
```
the BLOCK_EXCESSIVE bit doesn't change so I don't think we need the lock, and even if it did change, worst case we relay an excessive block...

ptschip replied:
we still need the lock for mi->IsValid() further up, but I can reduce the scope of the lock.

ptschip committed:
Reduce the scope of READLOCK(cs_mapBlockIndex)

ptschip committed:
fHaveData as we do not really need a lock here
